Rebased 191/712 implementation

holiman opened the pull request:
This PR is work done by @PaulRBerg,  I rebased https://github.com/ethereum/go-ethereum/pull/17789 on latest master. I didn't push it to the original PR, because I'm not 100% sure I didn't screw up anything during the rebase. 

I wanted it rebased in order to review it properly.

PaulRBerg commented:
Hey Martin, many thanks for this. Seeing the CI tests failed, I made a fix for it, but I can't seem to be able to push into this PR.

It's just a small change, a reorder of the imports: https://github.com/PaulRBerg/go-ethereum/blob/rebased_191/signer/core/signed_data.go#L32

gballet commented on signer/core/signed_data.go:
```diff
@@ -0,0 +1,788 @@
+		bigIntValue, _ := encValue.(*big.Int)
+		return fmt.Sprintf("%d (0x%x)", bigIntValue, bigIntValue)
+	}
+	return "NA"
```
let's return an error here: when it's called by `SignTypedData` the user will want to know why their request is not accepted.

gballet commented on signer/core/signed_data_test.go:
```diff
@@ -0,0 +1,715 @@
+	//formatted := d.FormatData(d.PrimaryType,d.Message)
+	//for _,item := range formatted{
+	//	fmt.Printf("%v\n", item.Pprint(0))
+	//}
```
don't forget to remove this before merging

gballet commented on signer/core/api.go:
```diff
@@ -170,11 +173,13 @@ type (
+		ContentType string                  `json:"content_type"`
+		Address     common.MixedcaseAddress `json:"address"`
+		Rawdata     interface{}             `json:"raw_data"`
+		//Message     string                  `json:"message"`
```
don't forget to remove this before merging

gballet commented on signer/core/signed_data.go:
```diff
@@ -0,0 +1,788 @@
+	return buffer.Bytes()
+}
+
+func (typedData *TypedData) TypeHash(primaryType string) hexutil.Bytes {
```
Public functions need to have a comment

gballet commented on signer/core/signed_data.go:
```diff
@@ -0,0 +1,788 @@
+}
+
+// Map is a helper function to generate a map version of the domain
+func (domain *TypedDataDomain) Map() map[string]interface{} {
```
The spec says that domain descriptions could be extended in the future. I would prefer something that copies (if necessary) the data from the map without making too many assumptions about its content.

gballet commented on signer/core/signed_data_test.go:
```diff
@@ -0,0 +1,715 @@
+	}
+}
+
+func TestMalformedData1(t *testing.T) {
```
please describe in what respect the data is malformed, so make it more explicit to the person for whom it will fail

gballet commented:
@holiman : I forgot in my review (and @PaulRBerg also noticed) that some of the headers got mixed up in the rebase).

holiman commented on signer/core/signed_data.go:
```diff
@@ -0,0 +1,788 @@
+
+		req = &SignDataRequest{ContentType: mediaType, Rawdata: plainData, Message: message, Hash: sighash}
+	default:
+		return nil, fmt.Errorf("content type '%s' not implemented for signing", contentType)
```
I think content-type is ok, the alternative is to have to implement new methods for different things. I do agree that we could default to `text/plain` though

holiman commented:
I addressed some of the concerns. I agree about the others, except that I think content-type is probably the best way forward. External libraries that interface with this api need to be rewritten, but if we keep adding new methods for different types of signing, we'll force them to have to be rewritten every time we add something.

holiman commented:
@PaulRBerg can you continue on the remaining comments from @gballet ? Since you can't push to this branch, you could either force-push this over your existing PR, or open a new one based off this (and I'll close this one if so)

PaulRBerg commented:
Of course, thank you so much for the help so far!! I was planning to make the fixes these days.

PaulRBerg commented:
I just force-pushed on #17789 the fixes as per @gballet's feedback. Some notes:

- My understanding is that you're suggesting to make the implementation of `Map` for the domain somehow more general-purpose. The problem is that in the context of EIP-712 there's an emphasis on specificity, so, if we're defaulting to `json.Unmarshal` or something alike, types get mismatched. For instance, an `address` is parsed as a `string`.
- The rationale for not including `Validate` in the `EncodeData` function in the first place was to avoid unnecessary computation. I see nonetheless that robustness should be prioritised.
- I kept reference and primitive types separate because their behaviour is different - a part I spent a lot of time to get right, as there are many variatons. For example:
    - Referenced types is part of `TypedData.Types` and any custom type must be in that list.
    - A primitive`bytes16` is totally different to `Bytezzz16`, because the latter is just a custom type, not a byte array with a maximum length of 16.
